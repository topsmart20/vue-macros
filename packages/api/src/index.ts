export * from '@vue-macros/common'

export * from './vue'
export * from './resolve'
export * from './ts'
export * from './utils'
