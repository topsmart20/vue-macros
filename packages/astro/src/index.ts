import { type AstroIntegration, type ViteUserConfig } from 'astro'
import { type Options, resolveOptions } from 'unplugin-vue-macros'
import { type Plugin } from 'vite'
import VueMacros from 'unplugin-vue-macros/vite'

export type VueMacrosOptions = Options

function findPluginAndRemove(
  name: string,
  plugins: ViteUserConfig['plugins']
): Plugin | undefined {
  const idx = plugins!.findIndex(
    (plugin) => plugin && 'name' in plugin && plugin.name === name
  )
  if (idx === -1) return
  const plugin = plugins![idx]
  plugins!.splice(idx, 1)
  return plugin as any
}

export default function (options?: VueMacrosOptions): AstroIntegration {
  return {
    name: '@vue-macros/astro',
    hooks: {
      'astro:config:setup': ({ config }) => {
        const resolvedOptions = resolveOptions(options ?? {})
        const vue = findPluginAndRemove('vite:vue', config.vite.plugins)
        const vueJsx = findPluginAndRemove('vite:vue-jsx', config.vite.plugins)

        config.vite.plugins?.push(
          VueMacros({
            ...resolvedOptions,
            plugins: {
              vue,
              vueJsx,
            },
          })
        )
      },
    },
  }
}
