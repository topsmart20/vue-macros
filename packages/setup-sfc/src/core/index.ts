import {
  MagicString,
  babelParse,
  getLang,
  getTransformResult,
} from '@vue-macros/common'
import type { HmrContext } from 'vite'

export const SETUP_SFC_REGEX = /\.setup\.[cm]?[jt]sx?$/

export const transfromSetupSFC = (code: string, id: string) => {
  const lang = getLang(id)
  const program = babelParse(code, lang)

  const s = new MagicString(code)
  for (const stmt of program.body) {
    if (stmt.type !== 'ExportDefaultDeclaration') continue

    if (
      !['FunctionExpression', 'ArrowFunctionExpression'].includes(
        stmt.declaration.type
      )
    )
      throw new SyntaxError(`Default export must be a function`)

    s.append(`defineOptions({
  render: ${s.sliceNode(stmt.declaration)}
});`)
    s.removeNode(stmt)
  }

  const attrs = `${lang ? ` lang="${lang}"` : ''}`
  s.prepend(`<script setup${attrs}>`)
  s.append(`</script>`)

  return getTransformResult(s, id)
}

export const hotUpdateSetupSFC = (
  { modules }: HmrContext,
  filter: (id: unknown) => boolean
) => {
  function isSubModule(id: string) {
    const [filename, query] = id.split('?')
    if (!query) return false
    if (!filter(filename)) return false
    return true
  }
  const affectedModules = modules.filter((mod) => mod.id && isSubModule(mod.id))
  return affectedModules
}
