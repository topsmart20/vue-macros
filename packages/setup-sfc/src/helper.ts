export function emitHelper(
  emitFn: (evt: string, ...args: any[]) => void,
  key: string,
  value: unknown
) {
  emitFn(key, value)
  return value
}
