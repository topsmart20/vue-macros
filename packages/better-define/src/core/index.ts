import { MagicString, getTransformResult, parseSFC } from '@vue-macros/common'
import { analyzeSFC } from '@vue-macros/api'
import type { TSEmits, TSProps } from '@vue-macros/api'

export async function transformBetterDefine(
  code: string,
  id: string,
  isProduction?: boolean
) {
  const s = new MagicString(code)
  const sfc = parseSFC(code, id)
  if (!sfc.scriptSetup) return

  const offset = sfc.scriptSetup.loc.start.offset
  const result = await analyzeSFC(s, sfc)
  if (result.props) {
    await processProps(result.props)
  }
  if (result.emits) {
    processEmits(result.emits)
  }

  return getTransformResult(s, id)

  async function processProps(props: TSProps) {
    const runtimeDefs = await props.getRuntimeDefinitions()

    const runtimeDecls = `{\n  ${Object.entries(runtimeDefs)
      .map(([key, { type, required, default: defaultDecl }]) => {
        const escapedKey = JSON.stringify(key)
        let defaultString = ''
        if (defaultDecl) {
          defaultString = defaultDecl('default')
        }
        if (!isProduction) {
          return `${escapedKey}: { type: ${toRuntimeTypeString(
            type
          )}, required: ${required}, ${defaultString} }`
        } else if (type.some((el) => el === 'Boolean' || el === 'Function')) {
          return `${escapedKey}: { type: ${toRuntimeTypeString(
            type
          )}, ${defaultString} }`
        } else {
          // production: checks are useless
          return `${escapedKey}: ${
            defaultString ? `{ ${defaultString} }` : 'null'
          }`
        }
      })
      .join(',\n  ')}\n}`

    let decl = runtimeDecls
    if (props.withDefaultsAst && !props.defaults) {
      // dynamic defaults
      decl = `_BD_mergeDefaults(${decl}, ${s.sliceNode(
        props.withDefaultsAst.arguments[1],
        { offset }
      )})`
      // add helper
      s.prependLeft(
        offset,
        `import { mergeDefaults as _BD_mergeDefaults } from 'vue'`
      )
    }
    decl = `defineProps(${decl})`

    s.overwriteNode(props.withDefaultsAst || props.definePropsAst, decl, {
      offset,
    })
  }

  function processEmits(emits: TSEmits) {
    const runtimeDecls = `[${Object.keys(emits.definitions)
      .map((name) => JSON.stringify(name))
      .join(', ')}]`
    s.overwriteNode(emits.defineEmitsAst, `defineEmits(${runtimeDecls})`, {
      offset,
    })
  }
}

function toRuntimeTypeString(types: string[]) {
  return types.length > 1 ? `[${types.join(', ')}]` : types[0]
}
