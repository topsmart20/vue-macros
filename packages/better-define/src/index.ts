import { createUnplugin } from 'unplugin'
import {
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  createFilter,
  detectVueVersion,
} from '@vue-macros/common'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import { transformBetterDefine } from './core'
import type { BaseOptions, MarkRequired } from '@vue-macros/common'
import type { UnpluginContextMeta } from 'unplugin'
import type { PluginContext } from 'rollup'

export interface Options extends BaseOptions {
  isProduction?: boolean
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework']
): OptionsResolved {
  const version = options.version || detectVueVersion()

  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC].concat(
      version === 2 && framework === 'webpack' ? REGEX_VUE_SUB : []
    ),
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = 'unplugin-vue-better-define'

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      transformInclude(id) {
        return filter(id)
      },

      async transform(code, id) {
        try {
          return await transformBetterDefine(code, id, options.isProduction)
        } catch (err: unknown) {
          this.warn(`${name} ${err}`)
          console.warn(err)
        }
      },

      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction
        },

        handleHotUpdate,
      },
    }
  }
)
