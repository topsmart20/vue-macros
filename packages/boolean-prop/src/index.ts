import { type Plugin } from 'vite'
import { type VuePluginApi, getVuePluginApi } from '@vue-macros/common'
import { generatePluginName } from '#macros' assert { type: 'macro' }
import { type Options, transformBooleanProp } from './core/index'

// legacy export
export * from './api'

const name = generatePluginName()

function rollup(options: Options = {}): Plugin {
  let api: VuePluginApi | null | undefined

  return {
    name,
    configResolved(config) {
      try {
        api = getVuePluginApi(config.plugins)
      } catch {}
    },
    buildStart(rollupOpts) {
      if (api === undefined)
        try {
          api = getVuePluginApi(rollupOpts.plugins)
        } catch (error: any) {
          this.warn(error)
          return
        }

      if (!api) return

      api.options.template ||= {}
      api.options.template.compilerOptions ||= {}
      api.options.template.compilerOptions.nodeTransforms ||= []

      api.options.template.compilerOptions.nodeTransforms.push(
        transformBooleanProp(options)
      )
    },
  }
}

export default {
  rollup,
  vite: rollup,
}
