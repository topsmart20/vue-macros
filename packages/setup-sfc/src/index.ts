import { createUnplugin } from 'unplugin'
import { createFilter } from '@rollup/pluginutils'
import { REGEX_SETUP_SFC } from '@vue-macros/common'
import { hotUpdateSetupSFC, transformSetupSFC } from './core'
import type { FilterPattern } from '@rollup/pluginutils'

export interface Options {
  include?: FilterPattern
  exclude?: FilterPattern
}

export type OptionsResolved = Omit<Required<Options>, 'exclude'> & {
  exclude?: FilterPattern
}

function resolveOption(options: Options): OptionsResolved {
  return {
    include: [REGEX_SETUP_SFC],
    ...options,
  }
}

const name = 'unplugin-vue-setup-sfc'

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}) => {
    const options = resolveOption(userOptions)
    const filter = createFilter(options.include, options.exclude)

    return {
      name,
      enforce: 'pre',

      transformInclude(id) {
        return filter(id)
      },

      transform(code, id) {
        try {
          return transformSetupSFC(code, id)
        } catch (err: unknown) {
          this.error(`${name} ${err}`)
        }
      },

      vite: {
        config() {
          return {
            esbuild: {
              exclude: options.include as any,
              include: options.exclude as any,
            },
          }
        },

        handleHotUpdate: (ctx) => {
          if (filter(ctx.file)) {
            return hotUpdateSetupSFC(ctx, filter)
          }
        },
      },
    }
  }
)
