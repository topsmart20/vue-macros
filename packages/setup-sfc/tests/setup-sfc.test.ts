/// <reference types="vite/client" />
import { resolve } from 'node:path'
import { describe, expect, test } from 'vitest'
import glob from 'fast-glob'
import esbuild from 'rollup-plugin-esbuild'
import Vue from 'unplugin-vue/vite'
import VueJsx from '@vitejs/plugin-vue-jsx'
import VueSetupSFC from '../src/rollup'
import { SETUP_SFC_REGEX } from '../src/core'
import { getCode } from './_utils'

describe('setup-component', async () => {
  test('isSetupSFC', () => {
    expect(SETUP_SFC_REGEX.test('foo.setup.ts')).toBe(true)
    expect(SETUP_SFC_REGEX.test('foo.setup.tsx')).toBe(true)
    expect(SETUP_SFC_REGEX.test('foo.setup.jsx')).toBe(true)
    expect(SETUP_SFC_REGEX.test('foo.setup.js')).toBe(true)
    expect(SETUP_SFC_REGEX.test('foo.setup.mjs')).toBe(true)
    expect(SETUP_SFC_REGEX.test('foo.setup.cts')).toBe(true)

    expect(SETUP_SFC_REGEX.test('foo.ts')).toBe(false)
  })

  describe('fixtures', async () => {
    const root = resolve(__dirname, '..')
    const files = await glob('tests/fixtures/*.{vue,[jt]s?(x)}', {
      cwd: root,
      onlyFiles: true,
    })

    for (const file of files) {
      test(file.replace(/\\/g, '/'), async () => {
        const filepath = resolve(root, file)

        const unpluginCode = await getCode(filepath, [
          VueSetupSFC(),
          Vue({
            include: [/\.setup\.[jt]sx?/],
          }) as any,
          VueJsx(),
          esbuild(),
        ])
        expect(unpluginCode).toMatchSnapshot()
      })
    }
  })
})
