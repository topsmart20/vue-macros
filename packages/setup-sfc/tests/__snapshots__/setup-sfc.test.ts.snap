// Vitest Snapshot v1

exports[`setup-component > fixtures > tests/fixtures/basic.setup.tsx 1`] = `
"import { defineComponent, createVNode } from 'vue';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"basic.setup\\",
  props: {
    foo: {
      type: String,
      required: true
    }
  },
  setup(__props) {
    defineOptions({
      render: () => {
        return createVNode(\\"div\\", null, null);
      }
    });
    return () => {
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var basic_setup = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { basic_setup as default };
"
`;

exports[`setup-component > fixtures > tests/fixtures/with-define-options.setup.tsx 1`] = `
"import { defineComponent, createVNode } from 'vue';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"with-define-options.setup\\",
  setup(__props) {
    defineOptions({
      name: \\"App\\"
    });
    defineOptions({
      render: () => {
        return createVNode(\\"div\\", null, null);
      }
    });
    return () => {
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var withDefineOptions_setup = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { withDefineOptions_setup as default };
"
`;
